File: Library/src/main/java/org/endeavourhealth/library/model/imq/Match.java
package org.endeavourhealth.library.model.imq;

import com.fasterxml.jackson.annotation.*;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

@JsonPropertyOrder({
  "notExists",
  "ifTrue",
  "ifFalse",
  "name",
  "description",
  "nodeRef",
  "header",
  "typeOf",
  "is",
  "path",
  "and",
  "or",
  "not",
  "where",
  "return",
  "then",
  ""
})
@JsonInclude(JsonInclude.Include.NON_DEFAULT)
public class Match extends IriLD implements HasPaths, Returnable {

  private Node graph;
  private Where where;
  private String description;
  private String nodeRef;
  private boolean optional;
  private Node typeOf;
  private String parameter;
  private String name;
  private List<Path> path;
  private FunctionClause function;
  private Entail entailment;
  private List<Return> returx;
  private RuleAction ifTrue;
  private RuleAction ifFalse;
  private boolean baseRule;
  private Integer ruleNumber;
  private boolean inverse;
  private boolean activeOnly;
  private List<Match> or;
  private List<Match> and;
  private List<Match> rule;
  private List<Match> any;
  private String libraryItem;
  private boolean invalid;
  private Node is;
  private List<GroupBy> groupBy;
  private String node;
  private OrderLimit orderBy;
  private String asDescription;
  private boolean notExists;
  private String errorMessage;
  private boolean draft;
  private Match then;
  private Having having;

  public List<Match> getAny() {
    return any;
  }

  public Match setAny(List<Match> any) {
    this.any = any;
    return this;
  }

  public Match addAny(Match any) {
    if (this.any == null) this.any = new ArrayList<>();
    this.any.add(any);
    return this;
  }

  public Match any(Consumer<Match> builder) {
    Match any = new Match();
    addAny(any);
    builder.accept(any);
    return this;
  }

  public Having getHaving() {
    return having;
  }

  public Match setHaving(Having having) {
    this.having = having;
    return this;
  }

  public Match having(Consumer<Having> builder) {
    Having having = new Having();
    setHaving(having);
    return this;
  }

  public Match getThen() {
    return then;
  }

  public Match setThen(Match then) {
    this.then = then;
    return this;
  }

  public Match then(Consumer<Match> builder) {
    this.then = new Match();
    builder.accept(this.then);
    return this;
  }

  public boolean isDraft() {
    return draft;
  }

  public Match setDraft(boolean draft) {
    this.draft = draft;
    return this;
  }

  public String getErrorMessage() {
    return errorMessage;
  }

  public Match setErrorMessage(String errorMessage) {
    this.errorMessage = errorMessage;
    return this;
  }

  public boolean isNotExists() {
    return notExists;
  }

  public Match rule(Consumer<Match> builder) {
    Match rule = new Match();
    addRule(rule);
    builder.accept(rule);
    return this;
  }

  public Match setNotExists(boolean notExists) {
    this.notExists = notExists;
    return this;
  }

  @JsonGetter
  public boolean notExists() {
    return notExists;
  }

  public Node getTypeOf() {
    return typeOf;
  }

  public Where getWhere() {
    return where;
  }

  @Override
  public String getDescription() {
    return description;
  }

  public String getParameter() {
    return parameter;
  }

  @Override
  public String getName() {
    return name;
  }

  @Override
  public List<Path> getPath() {
    return path;
  }

  public FunctionClause getFunction() {
    return function;
  }

  public Entail getEntailment() {
    return entailment;
  }

  public RuleAction getIfTrue() {
    return ifTrue;
  }

  public RuleAction getIfFalse() {
    return ifFalse;
  }

  public boolean isBaseRule() {
    return baseRule;
  }

  public Integer getRuleNumber() {
    return ruleNumber;
  }

  public void setRuleNumber(Integer ruleNumber) {
    this.ruleNumber = ruleNumber;
  }

  public boolean isInverse() {
    return inverse;
  }

  public List<Match> getOr() {
    return or;
  }

  public List<Match> getAnd() {
    return and;
  }

  public List<Match> getRule() {
    return rule;
  }

  public String getLibraryItem() {
    return libraryItem;
  }

  public boolean isInvalid() {
    return invalid;
  }

  public void setInvalid(boolean invalid) {
    this.invalid = invalid;
  }

  public Node getIs() {
    return is;
  }

  public String getAsDescription() {
    return asDescription;
  }

  public Match setActiveOnly(boolean activeOnly) {
    this.activeOnly = activeOnly;
    return this;
  }

  public Match setAsDescription(String asDescription) {
    this.asDescription = asDescription;
    return this;
  }

  public OrderLimit getOrderBy() {
    return orderBy;
  }

  public Match setOrderBy(OrderLimit orderBy) {
    this.orderBy = orderBy;
    return this;
  }

  public Match orderBy(Consumer<OrderLimit> builder) {
    this.orderBy = new OrderLimit();
    builder.accept(this.orderBy);
    return this;
  }

  public Match setNode(String node) {
    this.node = node;
    return this;
  }

  public List<GroupBy> getGroupBy() {
    return groupBy;
  }

  public Match setGroupBy(List<GroupBy> groupBy) {
    this.groupBy = groupBy;
    return this;
  }

  public Match addGroupBy(GroupBy group) {
    if (this.groupBy == null) this.groupBy = new ArrayList<>();
    this.groupBy.add(group);
    return this;
  }

  public Match groupBy(Consumer<GroupBy> builder) {
    GroupBy group = new GroupBy();
    addGroupBy(group);
    builder.accept(group);
    return this;
  }

  @JsonSetter
  public Match setIs(Node is) {
    this.is = is;
    return this;
  }

  public Match setLibraryItem(String libraryItem) {
    this.libraryItem = libraryItem;
    return this;
  }

  public Match setWhere(Where where) {
    this.where = where;
    return this;
  }

  public Match where(Consumer<Where> builder) {
    if (this.where == null) {
      Where where = new Where();
      setWhere(where);
      builder.accept(where);
    } else builder.accept(null);
    return this;
  }

  public Match setRule(List<Match> rule) {
    this.rule = rule;
    return this;
  }

  public Match addRule(Match rule) {
    if (this.rule == null) {
      this.rule = new ArrayList<>();
    }
    this.rule.add(rule);
    return this;
  }

  public Match setOr(List<Match> ors) {
    this.or = ors;
    return this;
  }

  public Match addOr(Match or) {
    if (this.or == null) {
      this.or = new ArrayList<>();
    }
    this.or.add(or);
    return this;
  }

  public Match or(Consumer<Match> builder) {
    Match or = new Match();
    addOr(or);
    builder.accept(or);
    return this;
  }

  public Match setAnd(List<Match> and) {
    this.and = and;
    return this;
  }

  public Match addAnd(Match and) {
    if (this.and == null) {
      this.and = new ArrayList<>();
    }
    this.and.add(and);
    return this;
  }

  public Match and(Consumer<Match> builder) {
    Match and = new Match();
    addAnd(and);
    builder.accept(and);
    return this;
  }

  public Match setBaseRule(boolean baseRule) {
    this.baseRule = baseRule;
    return this;
  }

  public Match setParameter(String parameter) {
    this.parameter = parameter;
    return this;
  }

  public Match setInverse(boolean inverse) {
    this.inverse = inverse;
    return this;
  }

  public Match setIfTrue(RuleAction ifTrue) {
    this.ifTrue = ifTrue;
    return this;
  }

  public Match setIfFalse(RuleAction ifFalse) {
    this.ifFalse = ifFalse;
    return this;
  }

  @JsonGetter
  public List<Return> getReturn() {
    return returx;
  }

  @JsonSetter
  public Match setReturn(List<Return> returns) {
    this.returx = returns;
    return this;
  }

  public Match addReturn(Return returnx) {
    if (this.returx == null) {
      this.returx = new ArrayList<>();
    }
    this.returx.add(returnx);
    return this;
  }

  public Match return_(Consumer<Return> builder) {
    Return returx = new Return();
    addReturn(returx);
    builder.accept(returx);
    return this;
  }

  public Match setEntailment(Entail entailment) {
    this.entailment = entailment;
    return this;
  }

  public Match setFunction(FunctionClause function) {
    this.function = function;
    return this;
  }

  public Match function(Consumer<FunctionClause> builder) {
    FunctionClause function = new FunctionClause();
    this.function = function;
    builder.accept(function);
    return this;
  }

  public Match setPath(List<Path> path) {
    this.path = path;
    return this;
  }

  public Match addPath(Path path) {
    if (this.path == null) {
      this.path = new ArrayList<>();
    }
    this.path.add(path);
    return this;
  }

  public Match path(Consumer<Path> builder) {
    Path path = new Path();
    this.addPath(path);
    builder.accept(path);
    return this;
  }

  @Override
  public Match setIri(String iri) {
    super.setIri(iri);
    return this;
  }

  @JsonIgnore
  public Match is(Consumer<Node> builder) {
    this.is = new Node();
    builder.accept(is);
    return this;
  }

  @JsonSetter
  public Match setTypeOf(Node typeOf) {
    this.typeOf = typeOf;
    return this;
  }

  public boolean isOptional() {
    return optional;
  }

  public Match setOptional(boolean optional) {
    this.optional = optional;
    return this;
  }

  public String getNodeRef() {
    return nodeRef;
  }

  public Match setNodeRef(String nodeRef) {
    this.nodeRef = nodeRef;
    return this;
  }

  public Node getGraph() {
    return graph;
  }

  public Match setGraph(Node graph) {
    this.graph = graph;
    return this;
  }

  public Match setTypeOf(String type) {
    this.typeOf = new Node().setIri(type);
    return this;
  }

  public Match setName(String name) {
    this.name = name;
    return this;
  }

  public Match setDescription(String description) {
    this.description = description;
    return this;
  }

  public String getNode() {
    return node;
  }

  public boolean isActiveOnly() {
    return activeOnly;
  }
}

File: src/interfaces/Match.ts
import z from "zod";

import { RuleAction } from "@/enums";
import { Entail } from "@/enums";

import { FunctionClause, FunctionClauseSchema } from "./FunctionClause";
import { GroupBy, GroupBySchema } from "./GroupBy";
import { Having, HavingSchema } from "./Having";
import { IriLD, IriLDSchema } from "./IriLD";
import { Node, NodeSchema } from "./Node";
import { OrderLimit, OrderLimitSchema } from "./OrderLimit";
import { Path, PathSchema } from "./Path";
import { Return, ReturnSchema } from "./Return";
import { Where, WhereSchema } from "./Where";

// export interface Match extends IriLD {
//   notExists?: boolean;
//   ifTrue?: RuleAction;
//   ifFalse?: RuleAction;
//   nodeRef?: string;
//   typeOf?: Node;
//   is?: Node;
//   and?: Match[];
//   or?: Match[];
//   where?: Where;
//   then?: Match;
//   graph?: Node;
//   optional?: boolean;
//   parameter?: string;
//   function?: FunctionClause;
//   entailment?: Entail;
//   baseRule?: boolean;
//   ruleNumber?: number;
//   inverse?: boolean;
//   activeOnly?: boolean;
//   rule?: Match[];
// any?: Match[];
//   libraryItem?: string;
//   invalid?: boolean;
//   groupBy?: GroupBy[];
//   orderBy?: OrderLimit;
//   asDescription?: string;
//   errorMessage?: string;
//   draft?: boolean;
//   having?: Having;
//   description?: string;
//   path?: Path[];
//   node?: string;
//   return?: Return[];
// }

export const MatchSchema = IriLDSchema.extend({
  notExists: z.boolean().default(false),
  ifTrue: z.enum(RuleAction).optional(),
  ifFalse: z.enum(RuleAction).optional(),
  nodeRef: z.string().optional(),
  get typeOf(): z.ZodOptional<typeof NodeSchema> {
    return NodeSchema.optional();
  },
  get is(): z.ZodOptional<typeof NodeSchema> {
    return NodeSchema.optional();
  },
  get and(): z.ZodPrefault<z.ZodArray<typeof MatchSchema>> {
    return z.array(MatchSchema).prefault([]);
  },
  get or(): z.ZodPrefault<z.ZodArray<typeof MatchSchema>> {
    return z.array(MatchSchema).prefault([]);
  },
  get where(): z.ZodOptional<typeof WhereSchema> {
    return WhereSchema.optional();
  },
  get then(): z.ZodOptional<typeof MatchSchema> {
    return MatchSchema.optional();
  },
  get graph(): z.ZodOptional<typeof NodeSchema> {
    return NodeSchema.optional();
  },
  optional: z.boolean().default(false),
  parameter: z.string(),
  function: FunctionClauseSchema.optional(),
  entailment: z.enum(Entail).optional(),
  baseRule: z.boolean().default(false),
  ruleNumber: z.number().optional(),
  inverse: z.boolean().default(false),
  activeOnly: z.boolean().default(false),
  get rule(): z.ZodPrefault<z.ZodArray<typeof MatchSchema>> {
    return z.array(MatchSchema).prefault([]);
  },
  get any(): z.ZodPrefault<z.ZodArray<typeof MatchSchema>> {
    return z.array(MatchSchema).prefault([]);
  },
  libraryItem: z.string().optional(),
  invalid: z.boolean().default(false),
  groupBy: z.array(GroupBySchema).optional(),
  orderBy: OrderLimitSchema.optional(),
  asDescription: z.string().optional(),
  errorMessage: z.string().optional(),
  draft: z.boolean().default(false),
  having: HavingSchema,
  description: z.string().optional(),
  path: z.array(PathSchema).optional(),
  node: z.string().optional(),
  return: z.array(ReturnSchema)
});

export type Match = z.output<typeof MatchSchema>;
